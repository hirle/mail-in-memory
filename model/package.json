{
  "version": "1.0.0",
  "license": "MIT",
  "name": "@mail-in-memory/model",
  "author": "hirle",
  "main": "dist/index.js",
  "typings": "dist/index.d.ts",
  "files": [
    "dist",
    "src"
  ],
  "engines": {
    "node": ">=12"
  },
  "scripts": {
    "build": "rollup --config",
    "test": "jest"
  },
  "devDependencies": {
    "jest": "^26.6.3",
    "rollup": "^2.35.1",
    "@rollup/plugin-typescript": "^8.1.0",
    "tslib": "^2.0.3",
    "typescript": "^4.1.3"
  },
  "prettier": {
    "printWidth": 80,
    "semi": true,
    "singleQuote": true,
    "trailingComma": "es5"
  },
  "module": "dist/model.esm.js"
}
