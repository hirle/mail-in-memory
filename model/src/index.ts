
export { default as Mail, MailInterface } from './Mail';
export { SocketMessages as SocketMessages } from './SocketMessages';