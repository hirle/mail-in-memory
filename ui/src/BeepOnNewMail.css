.beep-on-new-mail {
    flex: 1;
    padding: 12px 12px;
    margin: 2px;
    border: solid thin #111;
    text-align: left;
}