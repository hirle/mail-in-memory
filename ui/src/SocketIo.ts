import io from 'socket.io-client';

export type NotifyEvent = (data: any) => void;

export default class SocketIo {

    private socket?: SocketIOClient.Socket;
    private observers: Map<string, NotifyEvent[]>;

    constructor() {
        this.socket = undefined;
        this.observers = new Map<string, NotifyEvent[]>();
    }

    public startOn(path: string) {
        this.socket = io(path);
    }

    public stop() {
        // TODO code the stop
        console.log('not implemented yet');
    }

    public addObserver( event: string, notifyEvent: NotifyEvent ){
        if( ! this.observers.has(event)) {
            this.observers.set(event, []);
            this.socket?.on(event, this.notify.bind(this, event))
        }
        this.observers.get(event)!.push(notifyEvent);
    }

    public isConnected(): boolean {
        return !!this.socket && this.socket.connected;
    }

    public notify(event: string, data: any){
        if( this.observers.has(event) ) {
            this.observers.get(event)!.forEach( notifyEvent => notifyEvent(data));
        }
    }

}