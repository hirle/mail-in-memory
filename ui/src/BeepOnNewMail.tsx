import React from 'react';
import {SocketMessages} from '@mail-in-memory/model';
import SocketIo from './SocketIo';
import './BeepOnNewMail.css';

interface ComponentProps {
    socketIo: SocketIo
}

interface ComponentState {
    beep: boolean,
    lastEmailDate?: Date
}


export default class BeepOnNewMail extends React.Component<ComponentProps, ComponentState> {

    private audioRef: React.RefObject<HTMLAudioElement>;

    constructor(props: ComponentProps) {
        super(props);
        this.state = {beep: true};
        this.audioRef = React.createRef();
    }

    componentDidMount() {
        this.props.socketIo.addObserver(SocketMessages.NewMail, this.onNewMail.bind(this))
    }

    onNewMail() {
        this.setState({lastEmailDate: new Date()});
        console.log(this.state.beep);
        if( this.state.beep && this.audioRef.current) {
            this.audioRef.current.play()
                .catch(err => {
                    console.log(err);
                })
        }
    }

    handleCheckboxChange = ( event: React.ChangeEvent<HTMLInputElement> ) => this.setState({ beep: event.target.checked })

    render() {
      return (
        <div className="beep-on-new-mail"> 
            <audio ref={this.audioRef} src="beep.mp3"/>
            <label>
                <input type="checkbox" checked={this.state.beep} onChange={this.handleCheckboxChange}></input>
                <span>Beep on new Mail</span>
            </label>
            { this.state.lastEmailDate && <div>Last email received at: {this.state.lastEmailDate.toLocaleString()}</div>}
        </div>);
    }
  }