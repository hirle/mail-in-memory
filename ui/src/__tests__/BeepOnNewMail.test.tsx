import renderer from "react-test-renderer";
import SocketIo from "../SocketIo";
import BeepOnNewMail from "../BeepOnNewMail";
import { Mail } from "@mail-in-memory/model";
import { SocketMessages } from "@mail-in-memory/model";
import MockDate from 'mockdate';


describe("BeepOnNewMail", () => {
  
  afterEach( () => {
    MockDate.reset();
  });  
it("should update date on new mail", () => {
    const fixedDate = new Date('2021-02-28T09:33:30Z')
    MockDate.set(fixedDate);

    const mySocketIo = new SocketIo();
    const component = renderer.create(<BeepOnNewMail socketIo={mySocketIo} />);
    let initialTree = component.toJSON();
    expect(initialTree).toMatchInlineSnapshot(`
      <div
        className="beep-on-new-mail"
      >
        <audio
          src="beep.mp3"
        />
        <label>
          <input
            checked={true}
            onChange={[Function]}
            type="checkbox"
          />
          <span>
            Beep on new Mail
          </span>
        </label>
      </div>
    `);

    const incomingMail = Mail.create({
      messageId: "123-234-456-789",
      fromAddress: "foo@email.org",
      toAddress: "jest@test.mail-in-memory",
      subject: "Cool",
      body: "With jest, unit testing is so cool",
      mailTimestamp: new Date("2021-02-28T09:33:22Z"),
    });
    mySocketIo.notify(SocketMessages.NewMail, incomingMail);

    let afterMail = component.toJSON();
    // eslint-disable-next-line
    expect(afterMail).toMatchInlineSnapshot(`
      <div
        className="beep-on-new-mail"
      >
        <audio
          src="beep.mp3"
        />
        <label>
          <input
            checked={true}
            onChange={[Function]}
            type="checkbox"
          />
          <span>
            Beep on new Mail
          </span>
        </label>
        <div>
          Last email received at: 
          ${fixedDate.toLocaleString()}
        </div>
      </div>
    `);
  });
});
