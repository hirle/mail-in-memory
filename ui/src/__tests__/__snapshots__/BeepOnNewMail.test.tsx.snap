// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BeepOnNewMail should update date on new mail 1`] = `
<div
  className="beep-on-new-mail"
>
  <audio
    src="beep.mp3"
  />
  <label>
    <input
      checked={true}
      onChange={[Function]}
      type="checkbox"
    />
    <span>
      Beep on new Mail
    </span>
  </label>
</div>
`;

exports[`BeepOnNewMail should update date on new mail 2`] = `
<div
  className="beep-on-new-mail"
>
  <audio
    src="beep.mp3"
  />
  <label>
    <input
      checked={true}
      onChange={[Function]}
      type="checkbox"
    />
    <span>
      Beep on new Mail
    </span>
  </label>
  <div>
    Last email received at: 
    2021-2-28 10:33:30
  </div>
</div>
`;
